README.md
<div align="center">
    <img src="docs/assets/py_cui_logo_new.svg" alt="PyCUI" width=300px>
</div>

<div align="center">
    <p><strong><br/>A library for creating CUI/TUI interfaces with pre-built widgets in python.</strong></p>
    <p>Easily build text-based user interfaces on both Linux and Windows.</p>
    <a href="#">
        <img src="https://github.com/jwlodek/py_cui/workflows/tests/badge.svg?branch=master" alt="Automated Test+Build Status">
    </a>
    <a href="https://pepy.tech/project/py-cui">
        <img src="https://pepy.tech/badge/py-cui" alt="Download Counter">
    </a>
    <a href="https://badge.fury.io/py/py-cui">
        <img src="https://badge.fury.io/py/py-cui.svg" alt="PyPI Version">
    </a>
    <a href="#">
        <img src="https://img.shields.io/github/license/jwlodek/py_cui.svg" alt="License">
    </a>
</div>

---

# py_cui

`py_cui` is a python library meant to simplify writing console user interfaces in python. It relies on `curses` for terminal rendering, which is traditionally a unix-specific, however, you may use the [windows-curses](https://github.com/zephyrproject-rtos/windows-curses) module to run `py_cui` on windows.

The main advantage `py_cui` has over typical text-based user interface builders is that it relies on widgets and a grid layout manager like most traditional graphical user interfaces. You may define a grid size, and then drop predefined widgets onto it into specific grid locations. Widgets can also be stretched accross multiple grid rows and columns. If you've ever made a Tkinter GUI, you will feel right at home. In addition, `py_cui` has support for a rich collection of interactive popups such as menus, forms, file dialogs, and more.

If you would like to contribute, feel free to make an issue or pull request after reading through the `CONTRIBUTING.md` file.

## Installation

You may install `py_cui` via pip (use `pip3` if python 2 and 3 are installed side by side)

```bash
pip install py-cui
```

Make sure to run as root/sudo as required. Note that the library is titled `py_cui`, but the name of the pypi package is `py-cui`, though in most cases, installing with

```bash
pip install py_cui
```

should work OK.

If you would like to use `py_cui` directly from source, clone this repository with:

```bash
git clone https://github.com/jwlodek/py_cui
```

Then build/install with `pip`:

```bash
cd py_cui
pip install .
```

If you would like to try it out without installing, first make sure that `curses` is installed (`windows-curses` if on windows), and then copy one of the example files into the top directory. You may then simply run it with `python3`.

## Examples

Below are some screenshots and gifs of programs written with py_cui. Additional (simpler) examples are available in the `examples/` directory of this repository.

### py_cui_2048 - A command line version of the classic 2048 game

[py_cui_2048](https://github.com/jwlodek/py_cui_2048) is a demo of using `py_cui` to build simple CUI games and applications. The entire game was made in under 400 lines of code.

<p align="center">
    <img src="docs/assets/py2048-demo.gif">
</p>

### recoverpy - A TUI that allows for recovering lost or deleted files

[recoverpy](https://github.com/PabloLec/recoverpy) is a utility for recovering files on linux. It will search through every inode to find your requested file block.

<p align="center">
    <img src="docs/assets/recoverpy-demo.gif">
</p>

### pyautogit - A command line interface for managing git repositories

[pyautogit](https://github.com/jwlodek/pyautogit) is the first larger scale project written in `py_cui`. Feel free to use it as a guide to programming with `py_cui`.

<p align="center">
    <img src="docs/assets/pyautogit-demo.gif">
</p>

## Getting started

The easiest way to get set up for developing a full `py_cui` application is to use the available `cookiecutter` template. First, install the `cookiecutter` tool:

```bash
pip install cookiecutter
```

Once it is installed, navigate to the directory in which you would like your projects to live, and run:

```bash
cookiecutter https://github.com/jwlodek/py_cui_cookiecutter
```

This will clone the template, and show prompts for several variables, including project name and description, as well as some developer information.
Once it is done, you can enter your project directory, install it with pip, and run it, and you should be able to see a `Hello World` example:

```bash
cd myproject
pip install .
myproject
```

You are now ready to extend this template for your own application!

## Writing a PyCUI

Basic usage of `py_cui` starts with creating a PyCUI object, and specifiying it's grid size. Keep in mind that grid cell height and width will be measured in terminal characters, not pixels, so there is a lower limit on legal grid size, and heights will be smaller values than widths. Create this object with:

```Python
root = py_cui.PyCUI(7, 9)
```

The above line will create a UI with 7 rows and 9 columns. Then, add widgets with the different add commands:

```Python
label = root.add_label('Label Text', 0, 0)
button = root.add_button('Button Text', 1, 2, column_span=2, command=my_function)
...
```

Finally, start the CUI with

```Python
root.start()
```

`py_cui` has support for custom key bindings for both the overview mode and focused mode, popup windows and prompts, color rendering rules, and several useful widgets. For more details on writing `py_cui` based interfaces, be sure to check on the examples and the documentation [here](https://jwlodek.github.io/py_cui-docs).

## Using a PyCUI

There are some basic rules that apply to all `py_cui` based interfaces. There are three key operating modes - overview mode, focus mode, and popup mode.

### Overview Mode

Overview mode is the main control view of the interface. In this mode you use the arrow keys to move between widgets, and you may select widgets with the `Enter` key. By default, you may also press buttons in this mode, unless `auto_focus_buttons` is set to false when the CUI was created.

### Focus Mode

When in focus mode, you enter into a particular widget (For example a text box.). Each widget has some predefined basic controls, such as arrow keys to scroll in a `ScrollMenu`. You may also add keybindings to functions for each particular widget. These keybindings will only apply to a widget if it is in focus mode. Enter focus mode by navigating to a widget in overview mode and hitting the `Enter` key. Return to overview mode from focus mode by pressing `Escape`.

### Popup Mode

Popup mode simply displays a popup over the rest of the UI. Each popup type behaves slightly differently. For example, you may exit an info popup mode by pressing `Space`, `Enter`, or `Escape`, while a loading popup will remain onscreen until whatever operation is being run is terminated.

## Unit Tests

`py_cui` unit tests are written for `pytest`. Make sure `pytest` is installed, and simply run

```Bash
pytest
```

in the root directory to run all unit tests.

## Powered by py_cui

Below is a table of python projects developed with the help of `py_cui`. If you've made something, feel free to make a pull request to add it!

Project | Description
--------|-------------
[pypodcasts](https://github.com/jwlodek/pypodcasts) | A TUI for listening to and managing podcast feeds. (Coming Soon)
[pyautogit](https://github.com/jwlodek/pyautogit) | A command line UI for interfacing with git features for multiple repositories.
[recoverpy](https://github.com/PabloLec/recoverpy) | A CUI for recovering overwritten or deleted data on linux.
[py_cui_2048](https://github.com/jwlodek/py_cui_2048) | A CUI version of the classic 2048 game.
[unipkg](https://github.com/jwlodek/unipkg) | A CUI interface for managing all installed package managers on your system. (In-Progress)
[CUIAudioPlayer](https://github.com/jupiterbjy/CUIAudioPlayer) | A TUI for playing back audio files in a variety of formats.
[tinypub](https://github.com/HakierGrzonzo/tinyPub) | A console-based ebook reader.
[hue-tui](https://github.com/channel-42/hue-tui) | A CUI controlling for Philips Hue lights.
[mini-radio-player](https://github.com/wdog/mini-radio-player-pycui) | A terminal based radio interface for linux.

## License

BSD 3-Clause License

Copyright (c) 2019-2021, Jakub Wlodek


RELEASES.md
# py_cui Releases

This file will contain a changelog for all release versions of py_cui.

## v0.1.4 - Oct 29 2021

This release focuses on improving debugging tools for py_cui developers, adds mouse press support, improves user defineable callbacks for certain actions, and allows for "forgetting" widgets. It also drops support for python versions older than 3.6, improves some logging message levels, overhauls the widget ID system, and adds a large amount of various bugfixes and other minor new features. Please see the more detailed breakdown and issue list below.

**What this pull request changes**

* Live-Debug mode added. When enabled, a certain keybinding (default `Ctrl-D`) will open up an overlay widget with recent logger messages displayed. Log messages buffered up until a certain maximum. Standard logging levels are supported. Will make debugging UI issues and especially `py_cui` development much easier.
* Support for variety of mouse functions - left/right/middle clicks, double clicks, triple clicks, presses and releases
* Allow for user setting of mouse commands like key commands
* Allow setting an on-selection-change event function for scroll menus
* Python versions older than 3.6 dropped from support. I had wanted to keep around support for older versions due to the vintage of OS I used at work - but we have recently migrated everything to RHEL 8, giving me an excuse to drop older versions
* Some changes to log message levels
* Change to widget ID system - widgets now just get integers as IDs, this is cleaner
* Add ability to "forget" a widget and remove it from an active UI

### Issues Fixed

* #106 
* #58 
* #109 
* #98 
* #115 
* #113
* #100
* #72 
* #126 
* #119 
* #114 
* #107

### Breaking Changes

* There should be no major outward facing breaking changes with this release


## v0.1.3 - Mar 19 2021

This release adds filedialog popup support, a significant overhaul of the slider widget, overhaul of testing, scroll menu improvements, and minor bugfixes

### Features Added

* New popup - Filedialog popup allows for seamlessly asking user to select a file, directory, or a save location.
* Slider widget improved, adding options for alignment, as well as 
* Allow for border color to be set for in-focus and out of focus seperately
* Test overhaul to use pytest fixtures. Modernized pytest structure
* Version auto-detection during packaging process
* Example improvements + bugfixes
* IDE annotations added to examples
* Cyrillic input support for textboxes
* Avoid drawing bolded black on white - produces illegible results on many terminals
* Raise exceptions when function required to be implemented is not.

### Issues Fixed

* #49
* #68
* #75
* #57
* #61
* #86
* #64
* #88
* #90

### Breaking Changes

The slider widget has had some changes to the outer facing API. No other breaking changes were made.


## v0.1.2 - Sep 15, 2020

Next iterative release of `py_cui`. This release fixes some minor issues with the old version, adds form and slider popups and widgets, expands color and key options, and makes performance improvements.

### Features Added

* Huge performance improvement. Only redraw changed areas of screen.
* Functions to allow setting timeout for waiting for user input to refresh. Allows for editing values in second thread and seeing them update without having to interact with the UI
* Many new Colors added (56 total combinations, up from 10)
* Many new keys supported (Modifiers + letters, some special keys)
* Support for setting widget cycling keys for navigating between widgets even when in focus mode. (Defaults to Ctrl + left/right arrow keys).
* Support for a slider widget added.
* Form entry popup added. Allows for asking for several fields of input
* Greatly expanded coloration options. Can now individually set border color, text color, and selected text color

### Issues Fixed

* #54 
* #46 
* #50 
* #63
* #60 

### Breaking Changes

No breaking changes were made to the outer-facing API in this PR. Internally, certain functions had arguments added, but externally these are treated as keyword arguments.

## v0.1.1 - May 23, 2020

Next iterative release of `py_cui`, meant to add some requested functionality and improvements, particularly for scroll and checkbox menu widgets. Also adds mouse click support 

### Features Added

* Create `CheckBoxMenuImplementation` class
* Allow menu items to receive objects as well as strings
* Add additional default keys for faster menu navigation
* Mouse click support
* Fix bug with block label widget centering
* Fix some minor encapsulation issues

### Issues Fixed

* #44 
* #45 
* #16 
* #51 

### Breaking Changes

The only (minor) breaking change in this release is the change in the scoping of the title bar variable in the root PyCUI class. Most applications should not be affected, unless referencing the variable directly, instead of through getters/setters.

## v0.1.0 - Apr 18, 2020

First alpha release of `py_cui`, adds many requested features, but has some minor breaking changes, that shouldn't affect many programs.

### Features Added

* Add logging support for debug purposes
* Improve code reuse, restructure widget and popup classes
* Add set of central base classes for UI elements and implementations
* Move CI/CD to github actions
* Fix bug with window resize with widget sets on `win32`
* Improve error handling in main draw loop
* Improved widget navigation
* Fix `KEY_BACKSPACE` on MacOS
* Add ability to run in simulated terminal to improve testing capabilities

### Issues fixed

* #40 
* #38 
* #36 
* #35 
* #27 
* #21 

### Breaking Changes

* WidgetSet objects should no longer be directly created, but instead should be spawned with the `create_new_widget_set` function.
* Many internal variable name changes, are now accessible via getter and setter methods. Any instances where direct access to internal variables was made may be broken, but can easily be replaced without functionality loss with getter/setter alternatives

py_cui/keys.py
"""Module containing constants and helper functions for dealing with keys.
"""

# Author:    Jakub Wlodek
# Created:   12-Aug-2019


from sys import platform
import curses
from typing import Optional

# Some simple helper functions

def get_ascii_from_char(char: str) -> int:
    """Function that converts ascii code to character

    Parameters
    ----------
    char : character
        character to convert to ascii

    Returns
    -------
    ascii_code : int
        Ascii code of character
    """

    return ord(char)


def get_char_from_ascii(key_num: int) -> Optional[str]:
    """Function that converts a character to an ascii code

    Parameters
    ----------
    ascii_code : int
        Ascii code of character

    Returns
    -------
    char : character
        character converted from ascii
    """

    if key_num is None:
        return None

    return chr(key_num)


# Supported py_cui keys
KEY_ENTER       = get_ascii_from_char('\n')
# Escape character is ascii #27
KEY_ESCAPE      = 27
KEY_SPACE       = get_ascii_from_char(' ')
KEY_DELETE      = curses.KEY_DC
KEY_TAB         = get_ascii_from_char('\t')

KEY_BACKSPACE = [8, 127, curses.KEY_BACKSPACE]

SPECIAL_KEYS = [KEY_ENTER, KEY_ESCAPE, KEY_SPACE, KEY_DELETE, KEY_TAB] + KEY_BACKSPACE

# Page navigation keys
KEY_PAGE_UP     = curses.KEY_PPAGE
KEY_PAGE_DOWN   = curses.KEY_NPAGE
KEY_HOME        = curses.KEY_HOME
KEY_END         = curses.KEY_END

# Arrow Keys
KEY_UP_ARROW    = curses.KEY_UP
KEY_DOWN_ARROW  = curses.KEY_DOWN
KEY_LEFT_ARROW  = curses.KEY_LEFT
KEY_RIGHT_ARROW = curses.KEY_RIGHT


ARROW_KEYS = [KEY_UP_ARROW, KEY_DOWN_ARROW, KEY_LEFT_ARROW, KEY_RIGHT_ARROW]


if platform == 'linux' or platform == 'darwin':

    # Modified arrow keys
    KEY_SHIFT_LEFT  = 393
    KEY_SHIFT_RIGHT = 402
    KEY_SHIFT_UP    = 337
    KEY_SHIFT_DOWN  = 336

    KEY_CTRL_LEFT   = 560
    KEY_CTRL_RIGHT  = 545
    KEY_CTRL_UP     = 566
    KEY_CTRL_DOWN   = 525

elif platform == 'win32':

    KEY_SHIFT_LEFT  = 391
    KEY_SHIFT_RIGHT = 400
    KEY_SHIFT_UP    = 547
    KEY_SHIFT_DOWN  = 548

    KEY_CTRL_LEFT   = 443
    KEY_CTRL_RIGHT  = 444
    KEY_CTRL_UP     = 480
    KEY_CTRL_DOWN   = 481


MODIFIED_ARROW_KEYS = [KEY_SHIFT_LEFT, KEY_SHIFT_RIGHT, KEY_SHIFT_UP, KEY_SHIFT_DOWN,
                       KEY_CTRL_LEFT, KEY_CTRL_RIGHT, KEY_CTRL_UP, KEY_CTRL_DOWN]


# Standard letter keys
KEY_A_LOWER     = get_ascii_from_char('a')
KEY_B_LOWER     = get_ascii_from_char('b')
KEY_C_LOWER     = get_ascii_from_char('c')
KEY_D_LOWER     = get_ascii_from_char('d')
KEY_E_LOWER     = get_ascii_from_char('e')
KEY_F_LOWER     = get_ascii_from_char('f')
KEY_G_LOWER     = get_ascii_from_char('g')
KEY_H_LOWER     = get_ascii_from_char('h')
KEY_I_LOWER     = get_ascii_from_char('i')
KEY_J_LOWER     = get_ascii_from_char('j')
KEY_K_LOWER     = get_ascii_from_char('k')
KEY_L_LOWER     = get_ascii_from_char('l')
KEY_M_LOWER     = get_ascii_from_char('m')
KEY_N_LOWER     = get_ascii_from_char('n')
KEY_O_LOWER     = get_ascii_from_char('o')
KEY_P_LOWER     = get_ascii_from_char('p')
KEY_Q_LOWER     = get_ascii_from_char('q')
KEY_R_LOWER     = get_ascii_from_char('r')
KEY_S_LOWER     = get_ascii_from_char('s')
KEY_T_LOWER     = get_ascii_from_char('t')
KEY_U_LOWER     = get_ascii_from_char('u')
KEY_V_LOWER     = get_ascii_from_char('v')
KEY_W_LOWER     = get_ascii_from_char('w')
KEY_X_LOWER     = get_ascii_from_char('x')
KEY_Y_LOWER     = get_ascii_from_char('y')
KEY_Z_LOWER     = get_ascii_from_char('z')


# Shift modified (Uppercase) letters
KEY_A_UPPER     = get_ascii_from_char('A')
KEY_B_UPPER     = get_ascii_from_char('B')
KEY_C_UPPER     = get_ascii_from_char('C')
KEY_D_UPPER     = get_ascii_from_char('D')
KEY_E_UPPER     = get_ascii_from_char('E')
KEY_F_UPPER     = get_ascii_from_char('F')
KEY_G_UPPER     = get_ascii_from_char('G')
KEY_H_UPPER     = get_ascii_from_char('H')
KEY_I_UPPER     = get_ascii_from_char('I')
KEY_J_UPPER     = get_ascii_from_char('J')
KEY_K_UPPER     = get_ascii_from_char('K')
KEY_L_UPPER     = get_ascii_from_char('L')
KEY_M_UPPER     = get_ascii_from_char('M')
KEY_N_UPPER     = get_ascii_from_char('N')
KEY_O_UPPER     = get_ascii_from_char('O')
KEY_P_UPPER     = get_ascii_from_char('P')
KEY_Q_UPPER     = get_ascii_from_char('Q')
KEY_R_UPPER     = get_ascii_from_char('R')
KEY_S_UPPER     = get_ascii_from_char('S')
KEY_T_UPPER     = get_ascii_from_char('T')
KEY_U_UPPER     = get_ascii_from_char('U')
KEY_V_UPPER     = get_ascii_from_char('V')
KEY_W_UPPER     = get_ascii_from_char('W')
KEY_X_UPPER     = get_ascii_from_char('X')
KEY_Y_UPPER     = get_ascii_from_char('Y')
KEY_Z_UPPER     = get_ascii_from_char('Z')


LOWERCASE_LETTER_KEYS = list(range(97, 123))
UPPERCASE_LETTER_KEYS = list(range(65, 91))
LETTER_KEYS = UPPERCASE_LETTER_KEYS + LOWERCASE_LETTER_KEYS


# Control-modified Keys
KEY_CTRL_A      = 1
KEY_CTRL_B      = 2
KEY_CTRL_C      = 3
KEY_CTRL_D      = 4
KEY_CTRL_E      = 5
KEY_CTRL_F      = 6
KEY_CTRL_G      = 7
KEY_CTRL_H      = 8
KEY_CTRL_I      = 9
KEY_CTRL_J      = 10
KEY_CTRL_K      = 11
KEY_CTRL_L      = 12
KEY_CTRL_M      = 13
KEY_CTRL_N      = 14
KEY_CTRL_O      = 15
KEY_CTRL_P      = 16
KEY_CTRL_Q      = 17
KEY_CTRL_R      = 18
KEY_CTRL_S      = 19
KEY_CTRL_T      = 20
KEY_CTRL_U      = 21
KEY_CTRL_V      = 22
KEY_CTRL_W      = 23
KEY_CTRL_X      = 24
KEY_CTRL_Y      = 25
KEY_CTRL_Z      = 26

CTRL_MODIFIED_LETTERS = list(range(1,27))

_ALT_MODIFIER = 0
if platform == 'win32':
    _ALT_MODIFIER = 320

# Alt-modified keys
KEY_ALT_A       = 97 + _ALT_MODIFIER
KEY_ALT_B       = 98 + _ALT_MODIFIER
KEY_ALT_C       = 99 + _ALT_MODIFIER
KEY_ALT_D       = 100 + _ALT_MODIFIER
KEY_ALT_E       = 101 + _ALT_MODIFIER
KEY_ALT_F       = 102 + _ALT_MODIFIER
KEY_ALT_G       = 103 + _ALT_MODIFIER
KEY_ALT_H       = 104 + _ALT_MODIFIER
KEY_ALT_I       = 105 + _ALT_MODIFIER
KEY_ALT_J       = 106 + _ALT_MODIFIER
KEY_ALT_K       = 107 + _ALT_MODIFIER
KEY_ALT_L       = 108 + _ALT_MODIFIER
KEY_ALT_M       = 109 + _ALT_MODIFIER
KEY_ALT_N       = 110 + _ALT_MODIFIER
KEY_ALT_O       = 111 + _ALT_MODIFIER
KEY_ALT_P       = 112 + _ALT_MODIFIER
KEY_ALT_Q       = 113 + _ALT_MODIFIER
KEY_ALT_R       = 114 + _ALT_MODIFIER
KEY_ALT_S       = 115 + _ALT_MODIFIER
KEY_ALT_T       = 116 + _ALT_MODIFIER
KEY_ALT_U       = 117 + _ALT_MODIFIER
KEY_ALT_V       = 118 + _ALT_MODIFIER
KEY_ALT_W       = 119 + _ALT_MODIFIER
KEY_ALT_X       = 120 + _ALT_MODIFIER
KEY_ALT_Y       = 121 + _ALT_MODIFIER
KEY_ALT_Z       = 122 + _ALT_MODIFIER

ALT_MODIFIED_LETTERS = [_alt_mod_key + _ALT_MODIFIER for _alt_mod_key in range(97, 123)]

# List of all ctrl / alt modified key codes supported by py_cui
MODIFIED_LETTER_KEYS = CTRL_MODIFIED_LETTERS + ALT_MODIFIED_LETTERS

# Function keys
KEY_F0          = curses.KEY_F0
KEY_F1          = curses.KEY_F1
KEY_F2          = curses.KEY_F2
KEY_F3          = curses.KEY_F3
KEY_F4          = curses.KEY_F4
KEY_F5          = curses.KEY_F5
KEY_F6          = curses.KEY_F6
KEY_F7          = curses.KEY_F7
KEY_F8          = curses.KEY_F8
KEY_F9          = curses.KEY_F9
KEY_F10         = curses.KEY_F10
KEY_F11         = curses.KEY_F11
KEY_F12         = curses.KEY_F12

FUNCTION_KEYS = [KEY_F0, KEY_F1, KEY_F2, KEY_F3, KEY_F4, KEY_F5,
                 KEY_F6, KEY_F7, KEY_F8, KEY_F9, KEY_F10, KEY_F11, KEY_F12]


# Number keys
KEY_0 = get_ascii_from_char('0')
KEY_1 = get_ascii_from_char('1')
KEY_2 = get_ascii_from_char('2')
KEY_3 = get_ascii_from_char('3')
KEY_4 = get_ascii_from_char('4')
KEY_5 = get_ascii_from_char('5')
KEY_6 = get_ascii_from_char('6')
KEY_7 = get_ascii_from_char('7')
KEY_8 = get_ascii_from_char('8')
KEY_9 = get_ascii_from_char('9')

# Array storing number keys
NUMBER_KEYS = [KEY_0, KEY_1, KEY_2, KEY_3, KEY_4, KEY_5, KEY_6, KEY_7, KEY_8, KEY_9]
ALPHANUMERIC_KEYS = LETTER_KEYS + NUMBER_KEYS


# List of all key events supported by py_cui
ALL_KEYS = ALPHANUMERIC_KEYS + FUNCTION_KEYS + MODIFIED_LETTER_KEYS + SPECIAL_KEYS + ARROW_KEYS + MODIFIED_ARROW_KEYS


# Mouse click events
LEFT_MOUSE_CLICK        = curses.BUTTON1_CLICKED
LEFT_MOUSE_DBL_CLICK    = curses.BUTTON1_DOUBLE_CLICKED
LEFT_MOUSE_TRPL_CLICK   = curses.BUTTON1_TRIPLE_CLICKED
LEFT_MOUSE_PRESSED      = curses.BUTTON1_PRESSED
LEFT_MOUSE_RELEASED     = curses.BUTTON1_RELEASED

MIDDLE_MOUSE_CLICK      = curses.BUTTON2_CLICKED
MIDDLE_MOUSE_DBL_CLICK  = curses.BUTTON2_DOUBLE_CLICKED
MIDDLE_MOUSE_TRPL_CLICK = curses.BUTTON2_TRIPLE_CLICKED
MIDDLE_MOUSE_PRESSED    = curses.BUTTON2_PRESSED
MIDDLE_MOUSE_RELEASED   = curses.BUTTON2_RELEASED

RIGHT_MOUSE_CLICK        = curses.BUTTON3_CLICKED
RIGHT_MOUSE_DBL_CLICK    = curses.BUTTON3_DOUBLE_CLICKED
RIGHT_MOUSE_TRPL_CLICK   = curses.BUTTON3_TRIPLE_CLICKED
RIGHT_MOUSE_PRESSED      = curses.BUTTON3_PRESSED
RIGHT_MOUSE_RELEASED     = curses.BUTTON3_RELEASED

MOUSE_EVENTS = [LEFT_MOUSE_CLICK, LEFT_MOUSE_DBL_CLICK, LEFT_MOUSE_TRPL_CLICK, LEFT_MOUSE_PRESSED, LEFT_MOUSE_RELEASED,
                MIDDLE_MOUSE_CLICK, MIDDLE_MOUSE_DBL_CLICK, MIDDLE_MOUSE_TRPL_CLICK, MIDDLE_MOUSE_PRESSED, MIDDLE_MOUSE_RELEASED,
                RIGHT_MOUSE_CLICK, RIGHT_MOUSE_DBL_CLICK, RIGHT_MOUSE_TRPL_CLICK, RIGHT_MOUSE_PRESSED, RIGHT_MOUSE_RELEASED]


